tools/azsdk-cli/docs/specs/99-repo-tooling-contract.md

```diff
@@ -0,0 +1,322 @@
+
+By supporting `[Alias(...)]` script authors can easily coordinate changes to parameter names in `azsdk` without having to
+synchronize releases of the binary with script changes in the local repository.
+Aliases used by older `azsdk` versions can be kept around for backwards compatibility.
+
+To onboard an existing script to be used by the `azsdk` tool, all that needs to be done is add aliases for parameters.
+Alternatively, normalize the keys used by an `azsdk` invocation of the script to match its existing parameter names.
+
+```pwsh
+# eng/scripts/FooBar.ps1
+param(
+  [Alias('fooBarOption', 'foo-bar')]
+  [string] $FooBar
+)
+
+echo $FooBar
+```
+
+On the `azsdk` side, a service will be written to enable the following call patterns:
+
+- Within a helper method: a key/value parameters object can be created manually and passed to `Invoke`
+- In CLI mode: a key/value parameters object can be created from arguments and options, based on the parse result from `System.CommandLine`
+  - The key for arguments will be the key given in the CLI help text
+  - The key for options will be the non-aliased option flag value with leading hyphens trimmed, e.g. `--repo-path` becomes `repo-path`
+- In MCP mode: a key/value parameters object will be created from the variable names and values that are used to generate the tool's JSON schema
+
+```csharp
+public async Task<CLICheckResponse> FooBarCheck(string fooBarOption, CancellationToken ct = default)
+{
+    if (RepositoryService.HasImplementation("FooBarCheck"))
+    {
+        return RepositoryService.Invoke("FooBarCheck", new()
+        {
+            [nameof(fooBarOption)] = fooBarOption
+        });
+    }
+
+    // do fallback work here or throw not implemented
+}
+```
+
+```csharp
+// RepositoryService simple implementation
+public async Task<CLICheckResponse> Invoke(string commandName, OrderedDictionary<string, SomeParameterUnionType> params)
+{
+    var scriptPath = GetCommand(commandName);
+    var paramJson = JsonSerializer.Serialize(params, new JsonSerializerOptions { WriteIndented = false });
+    var command = $"$params = ('{paramJson}' | ConvertFrom-Json -AsHashtable); & {scriptPath} @params";
+    var options = new PowershellOptions(args: command);
+    var result = await _powershellHelper.Run(options, ct);
+    return new CLICheckResponse(result.ExitCode, result.Output);
+}
+```
+
+### User Experience
+
+[How will developers interact with this? Show examples of commands, outputs, or workflows]
+
+```bash
+# foobar command maps to eng/scripts/FooBar.ps1 in config
+# --foo-bar option maps to `foo-bar` key via helper method
+# The following powershell is executed by the repository service:
+#   $params = '{"foo-bar": "foobar"}' | ConvertFrom-Json -AsHashtable
+#   eng/scripts/FooBar.ps1 @params
+$ azsdk foobar --foo-bar foobar
```

Comment:
While PS supports splatting we could also implement our own version of splatting that would essentially output `--foo-bar "foobar"` which would enable other scripting language calls like npm or python directly. Those scripts already have to arg parse and can implement aliasing if they wish.